awx/main/tasks/callback.py

```diff
@@ -0,0 +1,203 @@
+import json
+import time
+import logging
+from collections import deque
+
+# Django
+from django.utils.timezone import now
+from django.conf import settings
+from django_guid.middleware import GuidMiddleware
+
+
+# AWX
+from awx.main.tasks.jobs import RunInventoryUpdate, RunProjectUpdate
+from awx.main.redact import UriCleaner
+from awx.main.constants import (
+    MINIMAL_EVENTS,
+)
+
+
+logger = logging.getLogger('awx.main.tasks.jobs')
+
+
+class RunnerCallback:
+    def __init__(self):
+        self.runner_callback.parent_workflow_job_id = None
+        self.runner_callback.host_map = {}
+        self.runner_callback.guid = GuidMiddleware.get_guid()
+        self.runner_callback.job_created = None
+        self.runner_callback.recent_event_timings = deque(maxlen=settings.MAX_WEBSOCKET_EVENT_RATE)
+
+    def event_handler(self, event_data):
+        #
+        # ⚠️  D-D-D-DANGER ZONE ⚠️
+        # This method is called once for *every event* emitted by Ansible
+        # Runner as a playbook runs.  That means that changes to the code in
+        # this method are _very_ likely to introduce performance regressions.
+        #
+        # Even if this function is made on average .05s slower, it can have
+        # devastating performance implications for playbooks that emit
+        # tens or hundreds of thousands of events.
+        #
+        # Proceed with caution!
+        #
+        """
+        Ansible runner puts a parent_uuid on each event, no matter what the type.
+        AWX only saves the parent_uuid if the event is for a Job.
+        """
+        # cache end_line locally for RunInventoryUpdate tasks
+        # which generate job events from two 'streams':
+        # ansible-inventory and the awx.main.commands.inventory_import
+        # logger
+        if isinstance(self, RunInventoryUpdate):
+            # Will removing this 'if'statement from here have any performance impact ?
+            self.end_line = event_data['end_line']
+
+        if event_data.get(self.event_data_key, None):
+            if self.event_data_key != 'job_id':
+                event_data.pop('parent_uuid', None)
+        if self.parent_workflow_job_id:
+            event_data['workflow_job_id'] = self.parent_workflow_job_id
+        event_data['job_created'] = self.job_created
+        if self.host_map:
+            host = event_data.get('event_data', {}).get('host', '').strip()
+            if host:
+                event_data['host_name'] = host
+                if host in self.host_map:
+                    event_data['host_id'] = self.host_map[host]
+            else:
+                event_data['host_name'] = ''
+                event_data['host_id'] = ''
+            if event_data.get('event') == 'playbook_on_stats':
+                event_data['host_map'] = self.host_map
+
+        if isinstance(self, RunProjectUpdate):
```

Comment:
@AlanCoding following up on https://github.com/ansible/awx/pull/11571/files#r795575348 , i can see a similar problem here. 
Can I do something like this : 
```python
 if isinstance(self,RunnercallbackForProjectUpdate)
```

This seems logical to me. I think it would work here.

Reply:
Yes, I agree.